### PR TITLE
New feature to clear only specific keys

### PR DESCRIPTION
**Pull request by Maschina**
This pull requests adds the feature to reset only single keys back to their default value. For optional keys, the default value will be `nil`.

### CODE REVIEW

File: Defaults.podspec [hunk 1]
```diff
@@ -1,6 +1,6 @@
 Pod::Spec.new do |s|
 	s.name = 'Defaults'
-	s.version = '2.0.2'
+	s.version = '2.1'
```
Comment (sindresorhus) [inline Defaults.podspec]: Don’t change the version. I will do so when publishing the new version.

File: Sources/Defaults/Defaults.swift [hunk 2]
```diff
@@ -65,6 +65,18 @@ public final class Defaults {
 			suite.removeObject(forKey: key)
 		}
 	}
+	
+	public func clear<T: Codable>(key: Defaults.Key<T>, suite: UserDefaults = .standard) {
+		for key in suite.dictionaryRepresentation().keys.filter({ $0 == key.name }) {
```
Comment (sindresorhus) [inline Sources/Defaults/Defaults.swift]: Use the “where” keyword instead of “filter”.

File: Sources/Defaults/Defaults.swift [hunk 3]
```diff
@@ -65,6 +65,18 @@ public final class Defaults {
 			suite.removeObject(forKey: key)
 		}
 	}
+	
+	public func clear<T: Codable>(key: Defaults.Key<T>, suite: UserDefaults = .standard) {
```
Comment (sindresorhus) [inline Sources/Defaults/Defaults.swift]: Can you add a documentation comment for this and the other clear method?

File: Sources/Defaults/Defaults.swift [hunk 4]
```diff
@@ -65,6 +65,18 @@ public final class Defaults {
...
+		for key in suite.dictionaryRepresentation().keys.filter({ $0 == key.name }) {
+			suite.removeObject(forKey: key)
+		}
+	}
+	
+	public func clear<T: Codable>(key: Defaults.OptionalKey<T>, suite: UserDefaults = .standard) {
```
Comment (sindresorhus) [inline Sources/Defaults/Defaults.swift]:
```suggestion
	public func clear<T: Codable>(_ key: Defaults.OptionalKey<T>, suite: UserDefaults = .standard) {
```

File: Defaults.podspec (hunk 1 above)
Comment (Maschina) [reply Defaults.podspec]:
```suggestion
	s.version = '2.0.2'
```

File: Sources/Defaults/Defaults.swift (hunk 2 above)
Comment (Maschina) [reply Sources/Defaults/Defaults.swift]: I have to apologize but I am not 100% sure what you mean. Were you thinking of `first(where: ...`?

File: Sources/Defaults/Defaults.swift (hunk 3 above)
Comment (sindresorhus) [reply Sources/Defaults/Defaults.swift]: I think it would be useful to accept multiple keys.

File: Sources/Defaults/Defaults.swift [hunk 5]
```diff
@@ -65,6 +65,18 @@ public final class Defaults {
 			suite.removeObject(forKey: key)
 		}
 	}
+	
+	public func clear<T: Codable>(key: Defaults.Key<T>, suite: UserDefaults = .standard) {
+		for key in suite.dictionaryRepresentation().keys.filter({ $0 == key.name }) {
+			suite.removeObject(forKey: key)
+		}
```
Comment (sindresorhus) [inline Sources/Defaults/Defaults.swift]: Why not just:

```swift
for key in keys {
	self[key] = key.defaultValue
}
```

?

File: Sources/Defaults/Defaults.swift (hunk 3 above)
Comment (Maschina) [reply Sources/Defaults/Defaults.swift]: Sure, documentation is useful here.

File: Configs/Defaults.plist [hunk 6]
```diff
@@ -13,7 +13,7 @@
 	<key>CFBundlePackageType</key>
 	<string>FMWK</string>
 	<key>CFBundleShortVersionString</key>
-	<string>2.0.2</string>
+	<string>$(MARKETING_VERSION)</string>
```
Comment (sindresorhus) [inline Configs/Defaults.plist]: Can you remove all the other unrelated changes like this one?

File: Sources/Defaults/Defaults.swift (hunk 3 above)
Comment (Maschina) [reply Sources/Defaults/Defaults.swift]: > I think it would be useful to accept multiple keys.

I agree, but I would rather implement it in an additional method.
Something like this: `public func clear<T: Codable>(keys: [Defaults.Key<T>]`

What's your opinion?

**Comment by sindresorhus** [comment]
This is what I currently use in one of my projects:

```swift
extension Defaults {
	/// Reset `Defaults.Key` items back to their default value.
	func reset<T: Codable>(_ keys: Defaults.Key<T>...) {
		reset(keys)
	}

	/// Reset `Defaults.Key` items back to their default value.
	func reset<T: Codable>(_ keys: [Defaults.Key<T>]) {
		for key in keys {
			self[key] = key.defaultValue
		}
	}

	/// Reset `Defaults.OptionalKey` items back to nil.
	func reset<T: Codable>(_ keys: Defaults.OptionalKey<T>...) {
		reset(keys)
	}

	/// Reset `Defaults.OptionalKey` items back to nil.
	func reset<T: Codable>(_ keys: [Defaults.OptionalKey<T>]) {
		for key in keys {
			self[key] = nil
		}
	}
}
```

File: Sources/Defaults/Defaults.swift (hunk 5 above)
Comment (Maschina) [reply Sources/Defaults/Defaults.swift]: Good idea. Would this also work for `OptionalKey`?

File: Configs/Defaults.plist (hunk 6 above)
Comment (Maschina) [reply Configs/Defaults.plist]:
```suggestion
	<string>2.0.2</string>
```

File: Sources/Defaults/Defaults.swift (hunk 5 above)
Comment (sindresorhus) [reply Sources/Defaults/Defaults.swift]: Yes: https://github.com/sindresorhus/Defaults/pull/21#issuecomment-519684656

**Comment by Maschina** [comment]
Nice code. Let me adapt it accordingly.

**Comment by sindresorhus** [comment]
I think I also prefer the `reset` name as it more accurately describe what it does; it resets back to the default value.

File: Sources/Defaults/Defaults.swift (hunk 3 above)
Comment (sindresorhus) [reply Sources/Defaults/Defaults.swift]: Do it like https://github.com/sindresorhus/Defaults/pull/21#issuecomment-519684656

File: Sources/Defaults/Defaults.swift (hunk 2 above)
Comment (sindresorhus) [reply Sources/Defaults/Defaults.swift]:
```swift
for suiteKey in suite.dictionaryRepresentation().keys where suiteKey == key {
```
